File: step1_start_project/src/test/java/migration/simple/controller/UserControllerTest.java
package migration.simple.controller;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import migration.simple.repository.UserRepository;
import migration.simple.responses.Response;
import migration.simple.responses.UserAddResponse;
import migration.simple.responses.UserResponse;
import migration.simple.types.User;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.http.*;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.web.client.RestClientException;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

import static org.junit.Assert.assertEquals;
import static org.mockito.BDDMockito.*;

@RunWith(SpringRunner.class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public class UserControllerTest {

    @Autowired
    private TestRestTemplate restTemplate;

    @Autowired
    private ObjectMapper objectMapper;

    @MockBean
    private UserRepository userRepositoryMock;

    @Test
    public void allUsersShouldBeReturnCorrectly() {
        List<User> users = Arrays.asList(
                new User(1L, "name1", "surname1", 25),
                new User(2L, "name2", "surname2", 30)
        );

        when(this.userRepositoryMock.findAllUsers()).thenReturn(users);
        UserResponse expectedResponse = new UserResponse(true, "return users", users);

        UserResponse actualResponse = restTemplate.getForObject("/users", UserResponse.class);

        assertEquals("invalid users response", expectedResponse, actualResponse);
    }

    @Test
    public void userShouldBeReturnCorrectly() {
        User user = new User(1L, "name1", "surname1", 25);
        when(this.userRepositoryMock.findUser(1L)).thenReturn(Optional.of(user));
        when(this.userRepositoryMock.findUser(2L)).thenReturn(Optional.empty());

        UserResponse expectedResponse = new UserResponse(true, "find user with requested id", Collections.singletonList(user));
        UserResponse actualResponse = restTemplate.getForObject("/user/{id}", UserResponse.class, 1L);
        assertEquals("invalid user find response", expectedResponse, actualResponse);

        UserResponse expectedMissedResponse = new UserResponse(false, "user not found", Collections.emptyList());
        UserResponse actualMissedResponse = restTemplate.getForObject("/user/{id}", UserResponse.class, 2L);
        assertEquals("invalid user not find response", expectedMissedResponse, actualMissedResponse);
    }

    @Test
    public void userShouldBeAddedCorrectly() throws JsonProcessingException {
        User newUser1 = new User(null, "name", "surname", 15);
        User newUser2 = new User(null, "name2", "surname2", 18);

        when(userRepositoryMock.addUser(newUser1)).thenReturn(Optional.of(15L));
        when(userRepositoryMock.addUser(newUser2)).thenReturn(Optional.empty());

        String request = objectMapper.writeValueAsString(newUser1);
        UserAddResponse expectedResponse = new UserAddResponse(true, "user add successfully", 15L);
        UserAddResponse actualResponse = putForObject("/user", request, UserAddResponse.class);
        assertEquals("invalid user add response", expectedResponse, actualResponse);
        verify(userRepositoryMock).addUser(eq(newUser1));

        String errorRequest = objectMapper.writeValueAsString(newUser2);
        UserAddResponse expectedErrorResponse = new UserAddResponse(false, "user not added", -1L);
        UserAddResponse actualErrorResponse = putForObject("/user", errorRequest, UserAddResponse.class);
        assertEquals("invalid user error add response", expectedErrorResponse, actualErrorResponse);
        verify(userRepositoryMock).addUser(eq(newUser2));
    }

    @Test
    public void userShouldBeDeletedCorrectly() {
        when(userRepositoryMock.deleteUser(1L)).thenReturn(true);
        when(userRepositoryMock.deleteUser(2L)).thenReturn(false);

        Response expectedResponse = new Response(true, "user has been deleted");
        Response actualResponse = deleteForObject("/user/{id}", Response.class, 1L);
        assertEquals("invalid delete response", expectedResponse, actualResponse);
        verify(userRepositoryMock).deleteUser(eq(1L));

        Response expectedErrorResponse = new Response(false, "user not been deleted");
        Response actualErrorResponse = deleteForObject("/user/{id}", Response.class, 2L);
        assertEquals("invalid delete error response", expectedErrorResponse, actualErrorResponse);
        verify(userRepositoryMock).deleteUser(eq(1L));
    }

    private <T> T putForObject(String url, String requestJson, Class<T> responseType, Object... uriVariables) throws RestClientException {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);

        final HttpEntity<String> httpEntity = new HttpEntity<>(requestJson, headers);
        final ResponseEntity<T> response = restTemplate.exchange(url, HttpMethod.PUT, httpEntity, responseType, uriVariables);
        final T result = response.getBody();
        return result;
    }

    private <T> T deleteForObject(String url, Class<T> responseType, Object... uriVariables) throws RestClientException {
        final HttpEntity<String> httpEntity = new HttpEntity<>("");
        final ResponseEntity<T> response = restTemplate.exchange(url, HttpMethod.DELETE, httpEntity, responseType, uriVariables);
        final T result = response.getBody();
        return result;
    }
}

File: step4_migration_to_webflux/src/test/kotlin/migration/simple/controller/UserControllerTest.kt
package migration.simple.controller

import com.fasterxml.jackson.databind.ObjectMapper
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.reset
import com.nhaarman.mockito_kotlin.whenever
import migration.simple.Application
import migration.simple.beans
import migration.simple.repository.UserRepository
import migration.simple.responses.DeleteResponse
import migration.simple.responses.UserAddResponse
import migration.simple.responses.UserResponse
import migration.simple.types.User
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.web.reactive.function.client.WebClient

@DisplayName("UserController test")
class UserControllerTest {
    private val userRepositoryMock = mock<UserRepository>()

    private val port = 8181
    private val configuration = beans().apply {
        bean { userRepositoryMock }
    }
    private var application = Application(port, configuration)

    private lateinit var client: WebClient

    private val objectMapper: ObjectMapper = ObjectMapper()

    @BeforeEach
    fun before() {
        reset(userRepositoryMock)
        application.start()
        client = WebClient.create("http://localhost:$port")
    }

    @AfterEach
    fun after() {
        application.stop()
    }

    @Test
    fun `all users should be return correctly`() {
        val users = listOf(
                User(1L, "name1", "surname1", 25),
                User(2L, "name2", "surname2", 30)
        )

        whenever(this.userRepositoryMock.findAllUsers()).thenReturn(users)
        val expectedResponse = UserResponse(true, "return users", users)

        client.get()
                .uri("/users")
                .retrieveResultForTest<UserResponse>()
                .expectNextMatches { it == expectedResponse }
                .verifyComplete()
    }

    @Test
    fun `user should be return correctly`() {
        val user = User(1L, "name1", "surname1", 25)
        whenever(this.userRepositoryMock.findUser(1L)).thenReturn(user)
        whenever(this.userRepositoryMock.findUser(2L)).thenReturn(null)

        val expectedResponse = UserResponse(true, "find user with requested id", listOf(user))

        client.get()
                .uri("/user/{id}", 1)
                .retrieveResultForTest<UserResponse>()
                .expectNextMatches { it == expectedResponse }
                .verifyComplete()

        val expectedMissedResponse = UserResponse(false, "user not found", emptyList())

        client.get()
                .uri("/user/{id}", 2)
                .retrieveResultForTest<UserResponse>()
                .expectNextMatches { it == expectedMissedResponse }
                .verifyComplete()
    }

    @Test
    fun `user should be added correctly`() {
        val newUser1 = User(null, "name", "surname", 15)
        val newUser2 = User(null, "name2", "surname2", 18)

        whenever(userRepositoryMock.addUser(newUser1)).thenReturn(15L)
        whenever(userRepositoryMock.addUser(newUser2)).thenReturn(null)

        val request = objectMapper.writeValueAsString(newUser1)
        val expectedResponse = UserAddResponse(true, "user add successfully", 15L)

        client.put()
                .uri("/user")
                .postBody(request)
                .retrieveResultForTest<UserAddResponse>()
                .expectNextMatches { it == expectedResponse }
                .verifyComplete()

        val errorRequest = objectMapper.writeValueAsString(newUser2)
        val expectedErrorResponse = UserAddResponse(false, "user not added", -1L)

        client.put()
                .uri("/user")
                .postBody(errorRequest)
                .retrieveResultForTest<UserAddResponse>()
                .expectNextMatches { it == expectedErrorResponse }
                .verifyComplete()
    }

    @Test
    fun `user should be deleted correctly`() {
        whenever(userRepositoryMock.deleteUser(1L)).thenReturn(true)
        whenever(userRepositoryMock.deleteUser(2L)).thenReturn(false)

        val expectedResponse = DeleteResponse(true, "user has been deleted")

        client.delete()
                .uri("/user/{id}", 1L)
                .retrieveResultForTest<DeleteResponse>()
                .expectNextMatches { it == expectedResponse }
                .verifyComplete()

        val expectedErrorResponse = DeleteResponse(false, "user not been deleted")

        client.delete()
                .uri("/user/{id}", 2L)
                .retrieveResultForTest<DeleteResponse>()
                .expectNextMatches { it == expectedErrorResponse }
                .verifyComplete()
    }


}